### PR TITLE
Adding Ruby LSP Support (Continuing PR-#11593)

### PR DESCRIPTION
**Pull request by billy1kaplan**
Trying to continue existing work on adding LSP backend to Ruby layer. Influenced heavily by the python lsp backend.

I'm not sure if it's best to start a new PR, but I wasn't sure how to continue in the existing one.

As far as I can tell, the open issues were:

1. Solve the key binding conflict with rubocop.
2. Figure out if documentation can be displayed without needing to execute yard gems.
3. Document it in the README.md.
4. Integrate with rvm.

Based on trying this out, I think #2 is crossed off since the documentation appeared to be working.

### CODE REVIEW

**Comment by billy1kaplan** [comment]
Based off of #11593

File: layers/+lang/ruby/funcs.el [hunk 1]
```diff
@@ -10,6 +10,56 @@
...

+;; lsp
+
+(defun spacemacs//ruby-setup-lsp ()
+  "Setup Ruby lsp."
+  (if (configuration-layer/layer-used-p 'lsp)
+      (lsp)
+    (message "`lsp' layer is not installed, please add `lsp' layer to your dotfile.")))
+
+(defun spacemacs//ruby-setup-lsp-company ()
```
Comment (yyoncho) [inline layers/+lang/ruby/funcs.el]: no longer needed, lsp-mode handles that automatically.

Comment (billy1kaplan) [reply layers/+lang/ruby/funcs.el]: To clarify, is this entire ruby-setup-lsp-company function not needed?

Comment (yyoncho) [reply layers/+lang/ruby/funcs.el]: Yes. lsp-mode automatically configures company-lsp if it is present.

**Comment by smile13241324** [comment]
@billy1kaplan would you mind squashing your commits? I would then go ahead and test this to get it merged.

**Comment by billy1kaplan** [comment]
@smile13241324 Sure, I should be able to tomorrow

**Comment by robbyoconnor** [comment]
I use rbenv... will this work?

**Comment by yyoncho** [comment]
@robbyoconnor I am not an expert in ruby but I think that dir local with a path to the solargraph server for the current rbenv will work. (lsp-mode does not have that variable though).

**Comment by billy1kaplan** [comment]
I'm also currently using rbenv. I just realized that this works when I launch emacs from my directory where I've done `gem install solargraph`, but doesn't seem to work when launched from elsewhere.

**Comment by robbyoconnor** [comment]
I've seen things fail miserably under rbenv -- for example, it can't seem to find executables

**Comment by robbyoconnor** [comment]
rvm introduces its own problems -- it caused zsh for segfault for me when trying to handle completions.

**Comment by billy1kaplan** [comment]
Any suggestions to move this forward? Is what you're describing specific to this change or is it more of an issue with the Ruby layer / Ruby environment managers?

**Comment by yyoncho** [comment]
You may add dap support as well, similar to the other lsp layers.

File: layers/+lang/ruby/funcs.el [hunk 2]
```diff
@@ -10,6 +10,49 @@
...
+(defun spacemacs//ruby-setup-lsp ()
+  "Setup Ruby lsp."
+  (if (configuration-layer/layer-used-p 'lsp)
+      (lsp)
+    (message "`lsp' layer is not installed, please add `lsp' layer to your dotfile."))
+(if (configuration-layer/layer-used-p 'dap)
+    (progn
+      (require 'dap-ruby)
+      (spacemacs/dap-bind-keys-for-mode 'ruby-mode))
+  (message "`dap' layer is not installed, please add `dap' layer to your dotfile.")))
```
Comment (robbyoconnor) [inline layers/+lang/ruby/funcs.el]: @yyoncho they did add dap support :)

Comment (yyoncho) [reply layers/+lang/ruby/funcs.el]: Thank you, I was very sleepy :)

Comment (robbyoconnor) [reply layers/+lang/ruby/funcs.el]: It's alright, no issue :)

**Comment by billy1kaplan** [comment]
Any additional feedback?

**Comment by smile13241324** [comment]
@billy1kaplan nope, not yet did not yet had the time to test this. I will have a look tomorrow.

File: layers/+lang/ruby/packages.el [hunk 3]
```diff
@@ -52,13 +52,13 @@
...
-  (with-eval-after-load 'company-dabbrev-code
-    (dolist (mode '(ruby-mode enh-ruby-mode))
-      (add-to-list 'company-dabbrev-code-modes mode))))
+  (add-hook 'ruby-mode-local-vars-hook #'spacemacs//ruby-setup-company)
+  (spacemacs|add-company-backends
+   :backends (company-files company-capf)
+   :modes ruby-mode
+   :variables
+   company-minimum-prefix-length 0
+   company-idle-delay 0.5))
```
[the remark this answers is not in the document]
Comment (billy1kaplan) [reply layers/+lang/ruby/packages.el]: Thanks for the feedback, I should have some time to look at the issues some more tomorrow.

[the remark this answers is not in the document]
Comment (billy1kaplan) [reply layers/+lang/ruby/packages.el]: >     * Please add a description and sample of how to install the supported LSP server for ruby. I have found out that lsp mode supports solargraph using gem install solargraph but this should be described in detail in the installation section of the layer readme.
> 
>     * Same for the packages in prerequisites here you should also mention which of these are necessary for the LSP backend and which are not.
> 
>     * ruby-backend should have 'robe as default value as to not break old dotfiles with the change.
Changes above are done.

>     * When I open a file test.rb I would have expected to have LSP loaded as I have the lsp layer in my dotfile, however this does not happen instead EnhRuby mode is loaded and the LSP keybindings are not in place.
I am having a bit of difficulty with this. When I open a Ruby file for the first time, I get the message "No LSP server for ruby-mode." If I close that same buffer and reopen the file, it works the second time. It also works the first time if I manually call `(lsp)` or launch spacemacs from the directory with the `gemfile` containing `solargraph`. Anyone have any suggestions on this?

>     * When I manually load lsp I would have expect autocomplete to work but I did not receive any autocompletion.
I've tried this out and it appears to be working now.

> 
> Please check the above mentioned points and mention me when I shall have another look.
> Btw as I am not very familiar with ruby would you mind adding a small file with sample code to test the feature?

Added a small file and uploaded a gif of the features in action.

[the remark this answers is not in the document]
Comment (yyoncho) [reply layers/+lang/ruby/packages.el]: > I am having a bit of difficulty with this. When I open a Ruby file for the first time, I get the message "No LSP server for ruby-mode."

You may check \*lsp-log\* for details why lsp-mode was not able to find ruby configuration.

[the remark this answers is not in the document]
Comment (billy1kaplan) [reply layers/+lang/ruby/packages.el]: Command "solargraph stdio" is not present on the path.

I'll look some more into this.

**Comment by billy1kaplan** [comment]
ruby-lsp-example

Sample of LSP in action

**Comment by robbyoconnor** [comment]
Great work!

**Comment by billy1kaplan** [comment]
Ok, I did a bit of digging. Here's what looks to be the issue for me on loading `solargraph`.

I'm currently on a Macbook. The first time I open a ruby file, lsp uses `exec-path` to search for solargraph. For some reason, the `/$HOME/.rbenv/bin` file is not in the path this time. Then, after I open a ruby file the exec-path is updated to include this file location. After that, I can open a `ruby` file and this time lsp finds solargraph and everything works. I haven't tried out other package managers, so I'm not sure what happens when using those.

Is this something weird specific to Mac? Maybe someone with more Emacs experience can point me in the right direction?

**Comment by robbyoconnor** [comment]
It's rbenv specific -- remember when I mentioned that rbenv was a bit
shitty to deal with? Welcome to the one of the shitty things.

**Comment by billy1kaplan** [comment]
Agreed that is a huge pain. I think it's because the `rbenv` initialization happens after the backend initialization. That means the first time it will fail without solargraph in the path but then will succeed after since `rbenv` has been initialized. I've added a hack to force the initialization before the backend. This seems to work, but please let me know if there's a better pattern for this.

**Comment by smile13241324** [comment]
@billy1kaplan I have checked your changes and looks good so far, shall I have another go?
Btw: I have noticed that the bindings for rubocop, have been changed from `r`to `R` can you explain why?

Unfortunately I have not yet a good idea how to fix the `rbenv` issue to be honest I would keep it like it is for now if it has no side effects.

**Comment by billy1kaplan** [comment]
@smile13241324 Yes, feel free to have another go. The rubocop `r` -> `R` was due to a conflicting keybinding with LSP. I'm sure there is probably a better way to handle this than changing the key binding.

As far as I can tell, this works for `rbenv` although it seems like the LSP takes a few seconds to start the first time opening a file.

File: layers/+lang/ruby/README.org [hunk 4]
```diff
@@ -48,14 +52,25 @@ This layer supports two different Ruby modes: Emacs’s built-in Ruby Mode and
...
-** Prerequisites
+** Backends
+*** Robe
+    The default Ruby-mode backend. See [[Prerequisites]] for the necessary Gems to support Robe.
+*** Language Server Protocal
+    IDE-like backend for ruby. See [[Prerequisites]] for the necessary Gems to support LSP. LSP also supports the DAP debugger which will be 
+
+    automatically configured upon using the LSP backend.
+    Dap-mode can be setup here: https://github.com/emacs-lsp/dap-mode#ruby
+    The
```
Comment (smile13241324) [inline layers/+lang/ruby/README.org]: - Please fix the indentation here, text under headlines should be uniformly formatted like it is done for the prerequisites headline. 
- Also you should add linebreaks to match a line limit of 80 chars in the documentation.
- Please also hide the external link in an org link expression.
- It looks like your text just stops at your last sentence, starting with `The`.

Comment (billy1kaplan) [reply layers/+lang/ruby/README.org]: Hope you had a good vacation! :)

I'm kind of an org-mode noob. I've tried to make the requested changes, but please let me know if they look ok. Thanks for the feedback!

**Comment by billy1kaplan** [comment]
@smile13241324 Do you get any messages in the log? Or is `(LSP)` not called at all?

**Comment by smile13241324** [comment]
@billy1kaplan sure. I have just checked my message file, there is no message in there indicating that lsp has been started. When I call `(LSP)` interactively it just works.

Here is the content of `*Log-LSP*` after executing `(LSP)`.

In my system solargraph is globally installed not the member of a local gem file.

I try to open a single ruby file. Do you have the same issue? Could really just my environment.
Here is the report for my system:
#### System Info :computer:
- OS: gnu/linux
- Emacs: 26.2
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. bf75b9d89)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(lsp dap
     (helm :variables helm-enable-auto-resize t)
     nim spacemacs-purpose ocaml
     (yang :variables yang-pyang-rules "ietf")
     ietf scheme
     (auto-completion :variables auto-completion-return-key-behavior 'complete auto-completion-tab-key-behavior 'cycle auto-completion-complete-with-key-sequence nil auto-completion-complete-with-key-sequence-delay 0.1 auto-completion-idle-delay nil auto-completion-private-snippets-directory nil auto-completion-enable-snippets-in-popup nil auto-completion-enable-help-tooltip t auto-completion-enable-sort-by-usage t)
     better-defaults
     (clojure :variables clojure-enable-fancify-symbols t clojure-enable-sayid t clojure-enable-clj-refactor t)
     gpu graphviz
     (plantuml :variables plantuml-jar-path "~/.plantuml/plantUml.jar" org-plantuml-jar-path "~/.plantuml/plantUml.jar")
     salt ibuffer emacs-lisp pdf
     (org :variables org-want-todo-bindings t org-enable-github-support t org-enable-bootstrap-support t org-enable-reveal-js-support t org-enable-hugo-support t org-enable-trello-support t org-enable-epub-support t org-enable-sticky-header nil)
     (shell :variables shell-enable-smart-eshell t shell-default-shell 'ansi-term shell-default-height 30 shell-default-position 'bottom shell-default-full-span nil shell-default-term-shell "/usr/bin/fish")
     syntax-checking version-control web-beautify semantic templates epub themes-megapack
     (c-c++ :variables c-c++-default-mode-for-headers 'c++-mode c-c++-enable-clang-support t c-c++-enable-clang-format-on-save t c-c++-enable-auto-newline t)
     (cmake :variables cmake-enable-cmake-ide-support t)
     git github sml bm
     (copy-as-format :variables copy-as-format-default "markdown" copy-as-format-asciidoc-include-file-name t)
     pass common-lisp
     (python :variables python-backend 'lsp python-test-runner 'pytest python-formatter 'yapf python-format-on-save t python-save-before-test t python-sort-imports-on-save t)
     windows-scripts agda jr shell-scripts prodigy emoji
     (markdown :variables markdown-live-preview-engine 'vmd)
     bibtex
     (json :variables json-fmt-tool 'web-beautify)
     debug
     (latex :variables latex-enable-auto-fill t latex-enable-magic nil latex-enable-folding t)
     lua html
     (javascript :variables javascript-import-tool 'import-js javascript-backend 'lsp javascript-fmt-tool 'web-beautify js2-basic-offset 2 js-indent-level 2)
     react yaml prettier
     (multiple-cursors :variables multiple-cursors-backend 'evil-mc)
     nginx docker csv search-engine
     (restclient :variables restclient-use-org t)
     restructuredtext selectric octave purescript speed-reading systemd imenu-list dap
     (java :variables java-backend 'lsp)
     groovy
     (go :variables go-backend 'lsp go-use-golangci-lint t gofmt-command "goimports" go-tab-width 4 go-format-before-save t go-use-gocheck-for-testing t go-use-test-args "-race -timeout 10s" godoc-at-point-function 'godoc-gogetdoc)
     coq django
     (elm :variables elm-sort-imports-on-save t elm-format-on-save t)
     idris
     (haskell :variables haskell-completion-backend 'lsp haskell-enable-hindent t haskell-process-type 'stack-ghci)
     pandoc sphinx parinfer asciidoc neotree ansible puppet rebox rust hy xkcd typography vimscript
     (terraform :variables terraform-auto-format-on-save t)
     (ruby :variables ruby-enable-enh-ruby-mode t ruby-backend 'lsp)
     ruby-on-rails asm
     (sql :variables sql-capitalize-keywords t)
     perl6 autohotkey
     (elixir :variables flycheck-elixir-credo-strict t)
     faust vagrant erlang
     (dash :variables helm-dash-docset-newpath "~/.local/share/Zeal/Zeal/docsets")
     games php nginx racket
     (colors :variables colors-colorize-identifiers 'all)
     d
     (typescript :variables typescript-backend 'lsp typescript-fmt-tool 'lsp typescript-fmt-on-save t))
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GSETTINGS GLIB NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD LCMS2

**Comment by smile13241324** [comment]
If you cannot reproduce my error I would merge this PR anyway as it is already opened for a long time and the feature is basically working, having to launch LSP manually is just a small issue and no reason to delay this any further.

**Comment by billy1kaplan** [comment]
I was able to reproduce the issue. The problem was with `enh-ruby-mode` enabled, then `ruby-mode` (where the lsp initialization logic was hooked into) wasn't called. Thank you for catching that.

File: layers/+lang/ruby/README.org [hunk 5]
```diff
@@ -50,14 +54,24 @@ This layer supports two different Ruby modes: Emacs’s built-in Ruby Mode and
...
+
+** Prerequisites 
 Some of the advanced features supported by this layer depend on external gems
 that need to be installed in the context of your project (see below for guidance
 based on your version manager):
 - =pry= and =pry-doc= are required for *jump to definition* and *code documentation* (=robe-mode=)
 - =ruby_parser= is required for *goto-step_definition* in =feature-mode=
 - =rubocop= is required for rubocop integration
 - =seeing_is_believing= helps you evaluate code inline
+- =solargraph= is required for using the lsp in =ruby-mode=
```
Comment (robbyoconnor) [inline layers/+lang/ruby/README.org]: Adding commands in a code block would help a lot.

**Comment by smile13241324** [comment]
@billy1kaplan have you seen the comment from @robbyoconnor? Could you add a small gem install command in the readme?

Shall I have another review?

**Comment by billy1kaplan** [comment]
@robbyoconnor I've added the command (`gem install solargraph`) into the Language Server Protocol Section. I'm hesitant to add one to the `Prerequisites` section due to the kind of catch all instructors at the end of the section.

**Comment by robbyoconnor** [comment]
However -- have a peek at the CI failures @billy1kaplan

**Comment by robbyoconnor** [comment]
@billy1kaplan you have trailing whitespace in your README -- do `SPC x d w` and save.

**Comment by smile13241324** [comment]
@billy1kaplan  @robbyoconnor is this ready to be reviewed now? I just had a quick look and for me it looks pretty good now.

**Comment by billy1kaplan** [comment]
@smile13241324 Yes, I think it is.

**Comment by robbyoconnor** [comment]
Feel free to merge :)

**Comment by smile13241324** [comment]
Thank you for contributing to Spacemacs! :+1: 
The PR has been cherry-picked into the develop branch, you can safely delete your branch.

**Pull request closed without merging**